### PR TITLE
refactoring tag based integration tests

### PR DESCRIPTION
**Pull request by individual-it**
## Description
This refactoring makes it easier to reuse test code across different tests suits.

## Related Issue
https://github.com/owncloud/QA/issues/467

## Motivation and Context
There is good code in the integration tests that would be very useful also for UI tests but its hard to reuse it. This is a proposal for a way forward to refactor code piece by piece to make it more reusable.

I've made the new helper functions static. This means a lot of parameter has to be passed around but also makes the function being independent.

## How Has This Been Tested?
Run integration tests locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by phil-davis — APPROVED** [review]
Looking nice.
Good to see the repeated -4 magic at least reduced to a single place :)

**Comment by SergioBertolinSG** [comment]
This wasn't even agreed yet. Please hold until I review and @PVince81 agrees.

**Comment by PVince81** [comment]
fine by me :+1:

**Comment by SergioBertolinSG** [comment]
What about making a library with useful commands, which can be used from everywhere just importing it with composer?

Something like pyocclient which can interact with a server. Separated completely from the behat step definitions.

We can write this library, taking code from here and later adapt the integration test to use it. Also the ui test.

What do you think?    @DeepDiver1975 also can give advice.

**Comment by butonic** [comment]
@SergioBertolinSG and how would behat communicate with the library? REST? That would mean implementing an API which would be capable of exactly the same operations as an actual API for the apps. I would prefer implementing the right API then.

Furthermore, testing a remote instance would be way slower.

I see no added value. Please explain what you are trying to achieve with this?

**Comment by SergioBertolinSG** [comment]
It will communicate like it is doing now with guzzle for example, using the library and using its contents.
The goal is to have a library which can be used not only in the integration tests, but in the UI tests or even in other programs interacting with owncloud server.
Also following this way we can keep the current structure of the integration tests for the moment until this library is ready to use. Without changing how they run.

**Comment by individual-it** [comment]
the tests are working for me locally but a lot of do fail in jenkins with error 500
https://jenkins.owncloud.org/job/owncloud-core/job/core/job/PR-28246/3/testReport/
could anybody try to run this tests and see if they work

**Comment by individual-it** [comment]
this helpers can be easily used in UI tests

File: tests/integration/features/bootstrap/BasicStructure.php [hunk 1]
```diff
@@ -86,10 +93,12 @@ public function usingServer($server) {
 		$previousServer = $this->currentServer;
 		if ($server === 'LOCAL'){
 			$this->baseUrl = $this->localBaseUrl;
+			$this->baseUrlWithoutOCSAppendix = substr($this->baseUrl, 0, -4);
```
Comment (phil-davis) [inline tests/integration/features/bootstrap/BasicStructure.php]: This looks annoying, repeating the same line of code twice. Maybe this and the "return" can both go after the whole if/else?

File: tests/TestHelpers/TagsHelper.php [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	 * @param string $baseUrl
+	 * @param string $taggingUser
+	 * @param string $password
+	 * @param string $tagName
+	 * @param string $fileName
+	 * @param string $fileOwner
+	 * @param int $davPathVersionToUse (1|2)
+	 * @return \GuzzleHttp\Message\FutureResponse|\GuzzleHttp\Message\ResponseInterface|NULL
+	 */
+	public static function tag(
```
Comment (butonic) [inline tests/TestHelpers/TagsHelper.php]: @PVince81 static makes sense here. The *Helper classes are just collections of functions. Your opinion?

Comment (butonic) [inline tests/TestHelpers/TagsHelper.php]: A Trait also doesn't add value...

File: tests/integration/features/bootstrap/BasicStructure.php (hunk 1 above)
Comment (individual-it) [inline tests/integration/features/bootstrap/BasicStructure.php]: 👍 dried it up

File: tests/TestHelpers/TagsHelper.php (hunk 2 above)
Comment (PVince81) [inline tests/TestHelpers/TagsHelper.php]: ok

**Comment by PVince81** [comment]
I'm fine with this. Do all integration tests still pass ? (Jenkins crashed again)

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by individual-it** [comment]
Build 19 did pass: https://jenkins.owncloud.org/job/owncloud-core/job/core/job/PR-28246/19/
After that there were only one small change to DRY the code

**Comment by phil-davis** [comment]
@individual-it @SergioBertolinSG 
what should happen about backporting this refactoring to stable10?

e.g. https://github.com/owncloud/core/pull/28347 adds on top of this
and whatever new tests or changes to tests happen now in master that use this TestHelpers stuff will also have trouble backporting to stable10.
and it is nice if we can run as many tests as possible back in stable10 so that 10.0.* gets the benefit of it.

**Comment by SergioBertolinSG** [comment]
I still think we should allocate time to prepare a library and remove all this helpers.

If this gets backported, please don't go further than stable10, don't reach stable9.1.

**Comment by phil-davis** [comment]
Definitely not to 9, because there will be too many diffs anyway.
It we backport this stuff to stable10, then if we refactor it again, it will be easy to also backport every new change to stable10 (becasue the stable10 code will be nearly identical to what is in master)

**Comment by SergioBertolinSG** [comment]
Yeah, that's right. Backport to stable10 then.

**Comment by phil-davis** [comment]
Backport to stable10 in #28394

**Pull request merged**